Lib/test/test_cmd.py

```diff
@@ -289,6 +289,31 @@ def do_tab_completion_test(self, args):
         self.assertIn(b'ab_completion_test', output)
         self.assertIn(b'tab completion success', output)
 
+    def test_bang_completion_without_do_shell(self):
+        script = textwrap.dedent("""
+            import cmd
+            class simplecmd(cmd.Cmd):
+                def completedefault(self, text, line, begidx, endidx):
+                    return ["hello"]
+
+                def default(self, line):
+                    if line == "! hello":
+                        print('tab completion success')
+                    else:
+                        print('tab completion failure')
+                    return True
+
+            simplecmd().cmdloop()
+        """)
+
+        # '! h' and complete 'ello' to '! hello'
+        input = b"! h\t\n"
```

Comment:
We should do add check here for `!h\t\n` (without space) as well. No need for another method, just another input.

Reply:
That does need to be another method, because the assertions are:
```py
        self.assertIn(b'hello', output)
        self.assertIn(b'tab completion success', output)
```
and those would pass if _either_ of the two inputs succeeded, not only if both of them did.

Reply:
Or maybe we could use a subtest for this? I know how to avoid duplication with pytest, but I'm not really familiar with unittest. Does this work?
```py
        # '! h' or '!h' and complete 'ello' to 'hello'
        for input in [b"! h\t\n", b"!h\t\n"]:
            with self.subTest(input=input):
                output = run_pty(script, input)
                self.assertIn(b'hello', output)
                self.assertIn(b'tab completion success', output)
```
Seems to...

Reply:
I mean just use the same script and do two `run_pty` - I don't think you even need a subtest. It's a very small case.